### ISSUE TITLE
tilt failing to apply namespaces / kinds first

### ISSUE THREAD

**Issue by landism** [windmilleng/tilt#1844]
Zeid Marouf reported on Slack that when they `tilt up` on https://github.com/zmarouf/kustomize-tilt-example, it fails to create some of the resources because the namespaces don't yet exist.

When I tried to repro, I didn't get an error due to lack of namespaces, but I did get an error due to a missing CRD. I'm unsure whether that's related.
```
ERROR: kubectl apply:
stderr: Error from server (NotFound): error when creating "STDIN": the server could not find the requested resource (post natsclusters.nats.io)
```

**Comment by nicks** [comment]
:clap: :clap: :clap: for a great example repo

that error message means "oops, we tried to install a NatsCluster before installing a Nats operator, and the API server doesn't know how to operate it. This feels like another use-case for https://github.com/windmilleng/tilt/blob/master/ROADMAP.md#service-dependencies. Or alternatively, we could just be smarter about recognizing CRD operators and installing them before the CRD instances

**Comment by landism** [comment]
OK, two issues here:

### Namespaces

Tilt [used to have logic](https://github.com/windmilleng/tilt/pull/1233) to apply namespaces first, and apparently that was [accidentally removed in a refactor](https://github.com/windmilleng/tilt/pull/1585). That should be simple to readd.

### CRDs

Even if we fix Tilt so that it applies namespaces and CRDs first, we'll have a problem in this repo.

The CRD here is not created directly by Tilt.

nats-operator [specifies a ClusterRole](https://github.com/nats-io/nats-operator/blob/master/deploy/00-prereqs.yaml#L43) with permissions to create CRDs,
then [specifies a Deployment](https://github.com/nats-io/nats-operator/blob/master/deploy/10-deployment.yaml) that starts a container that ensures those CRDs exist.

This means that Tilt will apply all the uncategorized YAML, including [this NatsCluster](https://github.com/zmarouf/kustomize-tilt-example/blob/master/base/nats-cluster.yaml) before applying any of the workloads, so the CRD won't exist yet. I don't have a good answer for this at the moment.

**Comment by nicks** [comment]
i wonder if we should have the integration tests deploy namespaces as part of `tilt up`. this seems to break a lot.

**Comment by zmarouf** [comment]
The namespace issue seems fixed on 0.9.3. 
For the CRD part, the containers crash a few times (1 or 2 restarts) before eventually running. 
That's perfectly acceptable.

Have you considered including a waitFor, waitOn, dependsOn mechanism?
I imagine the goal is to keep things simple but there might be use cases for it.
E.g I deploy Bitnami's SealedSecrets. I need to seal my secrets with the kubeseal binary but I can't do so before the keys get generated.
Hackish way would be to deploy the SealedSecrets, use a local("sleep 10") to wait and then local("kubeseal ...") followed by k8s_resource(["sealed-secrets.yaml", "deploymend-with-secrets.yaml"])

**Comment by maiamcc** [comment]
hey @zmarouf,

I'm uh. confused why the namespace issue would be fixed on 0.9.3, but I'm glad! I expect both the namespace and CRD issues to be fixed by #1859, which will be going out in a release in the next few business days! (Or if you're really excited, you can [install from source](https://github.com/windmilleng/tilt/blob/master/DEVELOPING.md).)

> Have you considered including a waitFor, waitOn, dependsOn mechanism?
doe #1846 seem like what you're talking about?

**Comment by zmarouf** [comment]
Thank you @maiamcc! #1846 is what I'm talking about!
Feel free to close this ticket unless you need it for internal tracking.

N.B I've been using Tilt from source because I'm fairly impatient :D
